### ISSUE TITLE
Terminado backend

### ISSUE THREAD

**Issue by staticfloat** [sourcelair/xterm.js#262]
Hello there, I'm coming here from [term.js](https://github.com/chjj/term.js), and I'm a little confused as to how to get this working with my [terminado](https://github.com/takluyver/terminado) backend.  Is such a thing even possible, or do I need to rewrite my backend to use `node.js` instead of python?  I'm having trouble finding any documentation on how the backend works or what it's compatible with.

Thanks!

**Comment by waywaaard** [comment]
Basically, you just need to hook your streams/events from the websockets to the xterm. You can also use https://github.com/sourcelair/xterm.js/tree/master/addons/attach to let xterm.js automatically work with the websockets.

If you are looking for the old `terminal.pipe()` function, then you need to roll your own.

**Comment by parisk** [comment]
**tl;dr: yes, you can.**

Hi @staticfloat, xterm.js is totally agnostic over the "transport protocol" or the back-end technology so you can definitely use it with terminado.

If terminado exchanges a stream of bare terminal data you can use the attach add-on as @waywaaard proposed.

If terminado exchanges encapsulated data (e.g. `{"data": "a"}`), then you will have to program the front-end to send the appropriate data and "decode" the received data before dispatching them to xterm.

Not sure how terminado works though, so you will have to take a deeper look at this.

Did this help you at all?

**Comment by parisk** [comment]
@staticfloat did the above answer help you? Should I move forward with closing this?

**Comment by staticfloat** [comment]
Sorry I was slow about this guys, I don't have much time to play around with this, so my progress has been slow.

I did get this working, thank you all for your help!  I had to learn a little bit about Javascript to do so, but that's hardly the worst thing in the world.  :)

In case you're curious, the necessary pieces to talk to the Python `terminado` backend are [here](https://github.com/staticfloat/webmux).  Most importantly, I modified `attach.js` to do the wrapping/unwrapping of data ([here](https://github.com/staticfloat/webmux/blob/074b98f0e16420f6bd65163dea34d10c7ef17322/webmux_static/terminado_attach.js#L80), [here](https://github.com/staticfloat/webmux/blob/074b98f0e16420f6bd65163dea34d10c7ef17322/webmux_static/terminado_attach.js#L84) and [here](https://github.com/staticfloat/webmux/blob/074b98f0e16420f6bd65163dea34d10c7ef17322/webmux_static/terminado_attach.js#L69-L76)) just as you suggested @parisk, and then rearchitected the resizing code to something a [little more straightforward](https://github.com/staticfloat/webmux/blob/074b98f0e16420f6bd65163dea34d10c7ef17322/webmux_static/main.js).

All in all, it's a nice upgrade, thank you guys for this great piece of software.

**Comment by parisk** [comment]
Awesome 👍 ! Happy you sorted that out. The solution seems indeed straightforward.

**Comment by scopatz** [comment]
Any chance we could get the terminado_attach.js in the main repo? Or better yet a Python backend demo, too?  I spent a long time futzing with other options and just found this one. Thanks for your consideration!

**Comment by staticfloat** [comment]
@scopatz A terminado demo would be useful, I agree.  So I've [opened a PR](https://github.com/sourcelair/xterm.js/pull/303) to include one.

**Comment by parisk** [comment]
You can find my reply at https://github.com/sourcelair/xterm.js/pull/303#issuecomment-251609201. I think we can make this happen after some tweaks.

**Comment by scopatz** [comment]
Thanks!
